src/state/mod.rs

```diff
@@ -83,25 +187,71 @@ impl Default for Config {
 
 pub struct State {
     pub project_dirs: ProjectDirs,
-    pub config: ConfigWrapper<Config>,
-    pub profiles: ConfigWrapper<ModProfiles>,
+    pub config: ConfigWrapper<VersionAnnotatedConfig>,
+    pub mod_data: ConfigWrapper<VersionAnnotatedModData>,
     pub store: Arc<ModStore>,
 }
 
 impl State {
-    pub fn new() -> Result<Self> {
+    pub fn init() -> Result<Self> {
         let project_dirs = ProjectDirs::from("", "", "drg-mod-integration")
             .context("constructing project dirs")?;
         std::fs::create_dir_all(project_dirs.cache_dir())?;
         std::fs::create_dir_all(project_dirs.config_dir())?;
-        let config = ConfigWrapper::<Config>::new(project_dirs.config_dir().join("config.json"));
-        let profiles =
-            ConfigWrapper::<ModProfiles>::new(project_dirs.config_dir().join("profiles.json"));
+
+        let config_path = project_dirs.config_dir().join("config.json");
+
+        let config = if config_path.exists() {
+            let config: MaybeVersionedConfig = std::fs::read(&config_path)
+                .ok()
+                .and_then(|s| serde_json::from_slice(&s).ok())
+                .unwrap_or_default();
+            let config = match config {
+                MaybeVersionedConfig::Versioned(v) => v,
+                MaybeVersionedConfig::Legacy(legacy) => {
+                    VersionAnnotatedConfig::V0_0_0(Config_v0_0_0 {
+                        provider_parameters: legacy.provider_parameters,
+                        drg_pak_path: legacy.drg_pak_path,
+                    })
+                }
+            };
+            ConfigWrapper::<VersionAnnotatedConfig>::new(&config_path, config)
+        } else {
+            ConfigWrapper::<VersionAnnotatedConfig>::new(&config_path, Default::default())
+        };
+        config.save().unwrap();
+
+        let legacy_mod_profiles_path = project_dirs.config_dir().join("profiles.json");
+        let mod_data_path = project_dirs.config_dir().join("mod_data.json");
+        let mod_data: MaybeVersionedModData = if mod_data_path.exists() {
+            std::fs::read(&mod_data_path)
+                .ok()
+                .and_then(|s| serde_json::from_slice(&s).ok())
+                .unwrap_or_default()
+        } else if legacy_mod_profiles_path.exists() {
+            let mod_data = std::fs::read(&legacy_mod_profiles_path)
+                .ok()
+                .and_then(|s| serde_json::from_slice(&s).ok())
+                .unwrap_or_default();
```

Comment:
I was using `unwrap_or_default` to quickly iterate during development but at this point `Default::default()` should only be returned if there is no such file. Otherwise the error should be returned so the user's config doesn't get wiped if there is a logic error with the migration or if the user was editing the config by hand and made a syntax error. This goes for every instance of `unwrap_or_default` being used to silently discard errors when reading files.

Reply:
Added proper error handling for failure to read config, failure to deserialize as v0.0.0 or legacy in [63a17cb](https://github.com/trumank/drg-mod-integration/pull/42/commits/63a17cb561c182c0db8fe92b5ad6a17e0512d14c).